src/bash/plugin.c

```diff
@@ -0,0 +1,394 @@
+    while (*current_plugin_node != NULL) {
+        current_plugin_node = &((*current_plugin_node)->next);
+		
+#ifdef DEBUG
+		itrace("Plugin: founded next plugin node: %p\n", *current_plugin_node);
+#endif
+    }
+    
+    /* append new plugin to tail node */
+    *current_plugin_node = new_plugin_node;
+	
+#ifdef DEBUG
+    itrace("Plugin: append new plugin node %p to %p\n", new_plugin_node, current_plugin_node);
+#endif
+}
+
+
+/* Load plugin by plugin path */
+void
+try_load_plugin_by_path(const char *plugin_path)
+{
+    /* Plugin handle */
+    void *plugin_handle;
+    if ( (plugin_handle = dlopen(plugin_path, RTLD_LAZY)) == NULL) {
+#ifdef DEBUG
+        itrace("Plugin: can't load plugin %s: %s\n", plugin_path, dlerror());
+#endif
+        return;
+    }
+
+    /* Check if plugin support shell execve method */
+    on_shell_execve_t* plugin_on_shell_execve_handle = dlsym(plugin_handle, on_shell_execve_function_name);
+    if (dlerror() != NULL) {
+        dlclose(plugin_handle);
+
+#ifdef DEBUG
+        itrace("Plugin: can't find on_shell_execve function %s: %s\n", plugin_path, dlerror());
+#endif
+        return;
+    }
+    
+
+    /* Check if plugin support un-initialization method */
+    plugin_uninit_t* plugin_uninit_handle = dlsym(plugin_handle, plugin_uninit_function_name);
+    if (dlerror() != NULL) {
+        dlclose(plugin_handle);
+
+#ifdef DEBUG
+        itrace("Plugin: can't find plugin_uninit function %s: %s\n", plugin_path, dlerror());
+#endif
+        return;
+    }
+
+    /* Check if plugin support initialization method */
+    plugin_init_t* plugin_init_handle = dlsym(plugin_handle, plugin_init_function_name);
+    if (dlerror() != NULL) {
+        dlclose(plugin_handle);
+
+#ifdef DEBUG
+        itrace("Plugin: can't find plugin_init function %s: %s\n", plugin_path, dlerror());
+#endif
+        return;
+    }
+    else {
+        /* Initialize plugin */
+        plugin_init_handle();
+    }
+
+    /* Add plugin to plugin list */
+    append_plugin(plugin_handle,
+                    plugin_on_shell_execve_handle,
+                    plugin_init_handle,
+                    plugin_uninit_handle);
+    
+#ifdef DEBUG
+    itrace("Plugin: plugin %s loaded\n", plugin_path);
+#endif
+}
+
+/* Load plugin by config file */
+void
+load_plugin_by_config(const char *config_filename)
+{
+    FILE *config_file;
+    char buffer[256];
+
+    config_file = fopen(config_filename, "r");
+    if(config_file == NULL) {
+#ifdef DEBUG
+            itrace("Plugin: can't open plugin config file %s: %s\n", config_filename, strerror(errno));
+#endif
+        return;
+    }
+
+    while(fgets(buffer, sizeof buffer, config_file)) {
+        if(*buffer == '#' || isspace(*buffer)) {
+            /* ignore comments or white space. */
+            continue; 
+        }
+        
+        /* read to first whitespace. */
+        strtok(buffer, " \t\n\r\f"); 
+
+        if(!strncmp(buffer, "plugin=", 7)) {
```

Comment:
>7 [](http://example.com/codeflow?start=39&length=1)

Is it `strlen()`? If you really want to magic number, add some comment so future dev could possible change related string.

Reply:
Fixed